File: typescript/src/connector.ts
/**
 * Connector — SDK-direct cloud deploys (no Terraform).
 *
 * Mirrors `connector/connector.go` and Python `vxsdk.Connector`. Unlike `cloud`
 * (which provisions via server-side Terraform under /api/v2/tenant/provision/*),
 * connector targets the node's /api/v2/connector/* surface, which provisions by
 * calling cloud-provider SDKs/REST directly — no Terraform, no Bash, no state:
 *
 *   - EC2 VM provision / studio deploy (SCP+SSH) / terminate
 *   - S3 static-site deploy and bucket creation
 *   - GCP Cloud Run container deploy (Cloud Run Admin REST v2)
 *   - Route53 subdomain records
 *   - ALB attach with target group
 *
 * Every request carries `username` for the Vault-backed credential lookup the
 * node performs. When omitted it defaults to the client's authed username.
 */

import type { Transport } from './transport.js';

/** A decoded JSON object response. */
export type Result = Record<string, unknown>;

/** Provision a single EC2 instance (no file deploy). */
export interface ConnectorVMInput {
  instanceName: string;
  instanceType?: string;
  region?: string;
  os?: string;
  volumeSize?: number;
  keyPairName?: string;
  sessionId?: string;
  organization?: string;
  tags?: Record<string, string>;
  /** Workspace owner for the Vault credential lookup. Defaults to the client's authed username. */
  username?: string;
}

/** Provision an EC2 instance then push a studio project to it over SFTP. */
export interface ConnectorDeployVMInput extends ConnectorVMInput {
  sessionId: string;
  storagePath?: string;
  subdomain?: string;
  deployCommand?: string;
}

/** Create a single S3 bucket. */
export interface ConnectorBucketInput {
  bucketName: string;
  region?: string;
  public?: boolean;
  organization?: string;
  username?: string;
}

/** Deploy a studio project to S3 as a static website. */
export interface ConnectorStaticSiteInput {
  sessionId: string;
  storagePath?: string;
  bucketName?: string;
  region?: string;
  indexDocument?: string;
  errorDocument?: string;
  subdomain?: string;
  customDomain?: string;
  organization?: string;
  username?: string;
}

/** Deploy a container image to Google Cloud Run. */
export interface ConnectorCloudRunInput {
  serviceName: string;
  image: string;
  region?: string;
  port?: number;
  envVars?: Record<string, string>;
  memory?: string;
  cpu?: string;
  maxInstances?: number;
  subdomain?: string;
  allowPublic?: boolean;
  sessionId?: string;
  organization?: string;
  username?: string;
}

/** Create a Route53 DNS record. */
export interface ConnectorSubdomainInput {
  subdomain: string;
  baseDomain?: string;
  hostedZoneId?: string;
  targetIp?: string;
  targetCname?: string;
  recordType?: string;
  ttl?: number;
  organization?: string;
  username?: string;
}

/** Attach instances to an application load balancer (creates ALB + target group). */
export interface ConnectorLBInput {
  name: string;
  instanceIds: string[];
  port?: number;
  healthPath?: string;
  region?: string;
  subdomain?: string;
  vpcId?: string;
  subnetIds?: string[];
  organization?: string;
  username?: string;
}

export class Connector {
  constructor(private t: Transport, private getUsername: () => string) {}

  private u(username?: string): string {
    return (username && username.trim()) || this.getUsername();
  }

  private async post(path: string, body: Record<string, unknown>): Promise<Result> {
    return (await this.t.postJSON<Result>(path, body)).body ?? {};
  }

  /** Launch a single EC2 instance (no file deploy). POST /api/v2/connector/vm/provision */
  async provisionVm(input: ConnectorVMInput): Promise<Result> {
    if (!input.instanceName) throw new Error('connector.provisionVm: instanceName is required');
    return this.post('/api/v2/connector/vm/provision', {
      username: this.u(input.username),
      ...compact({
        organization: input.organization,
        session_id: input.sessionId,
        instance_name: input.instanceName,
        instance_type: input.instanceType,
        region: input.region,
        os: input.os,
        volume_size: input.volumeSize,
        key_pair_name: input.keyPairName,
        tags: input.tags,
      }),
    });
  }

  /** Provision an EC2 instance then push a studio project to it. POST /api/v2/connector/vm/studio/deploy */
  async deployToVm(input: ConnectorDeployVMInput): Promise<Result> {
    if (!input.sessionId) throw new Error('connector.deployToVm: sessionId is required');
    return this.post('/api/v2/connector/vm/studio/deploy', {
      username: this.u(input.username),
      ...compact({
        organization: input.organization,
        session_id: input.sessionId,
        storage_path: input.storagePath,
        instance_name: input.instanceName,
        instance_type: input.instanceType,
        region: input.region,
        os: input.os,
        volume_size: input.volumeSize,
        key_pair_name: input.keyPairName,
        subdomain: input.subdomain,
        deploy_command: input.deployCommand,
        tags: input.tags,
      }),
    });
  }

  /** Terminate an EC2 instance. POST /api/v2/connector/vm/terminate */
  async terminateVm(instanceId: string, organization?: string): Promise<Result> {
    if (!instanceId) throw new Error('connector.terminateVm: instanceId is required');
    return this.post('/api/v2/connector/vm/terminate', compact({
      username: this.u(),
      organization,
      instance_id: instanceId,
    }));
  }

  /** Create a single S3 bucket. POST /api/v2/connector/s3/bucket */
  async createBucket(input: ConnectorBucketInput): Promise<Result> {
    if (!input.bucketName) throw new Error('connector.createBucket: bucketName is required');
    return this.post('/api/v2/connector/s3/bucket', {
      username: this.u(input.username),
      ...compact({
        organization: input.organization,
        bucket_name: input.bucketName,
        region: input.region,
        public: input.public,
      }),
    });
  }

  /** Deploy a studio project to S3 as a static website. POST /api/v2/connector/s3/studio/deploy */
  async deployStaticSite(input: ConnectorStaticSiteInput): Promise<Result> {
    if (!input.sessionId) throw new Error('connector.deployStaticSite: sessionId is required');
    return this.post('/api/v2/connector/s3/studio/deploy', {
      username: this.u(input.username),
      ...compact({
        organization: input.organization,
        session_id: input.sessionId,
        storage_path: input.storagePath,
        bucket_name: input.bucketName,
        region: input.region,
        index_document: input.indexDocument,
        error_document: input.errorDocument,
        subdomain: input.subdomain,
        custom_domain: input.customDomain,
      }),
    });
  }

  /** Deploy a container image to Google Cloud Run. POST /api/v2/connector/gcr/studio/deploy */
  async deployCloudRun(input: ConnectorCloudRunInput): Promise<Result> {
    if (!input.serviceName) throw new Error('connector.deployCloudRun: serviceName is required');
    if (!input.image) throw new Error('connector.deployCloudRun: image is required');
    return this.post('/api/v2/connector/gcr/studio/deploy', {
      username: this.u(input.username),
      ...compact({
        organization: input.organization,
        session_id: input.sessionId,
        service_name: input.serviceName,
        region: input.region,
        image: input.image,
        port: input.port,
        env_vars: input.envVars,
        memory: input.memory,
        cpu: input.cpu,
        max_instances: input.maxInstances,
        subdomain: input.subdomain,
        allow_public: input.allowPublic,
      }),
    });
  }

  /** Create a Route53 DNS record. POST /api/v2/connector/dns/subdomain */
  async createSubdomain(input: ConnectorSubdomainInput): Promise<Result> {
    if (!input.subdomain) throw new Error('connector.createSubdomain: subdomain is required');
    return this.post('/api/v2/connector/dns/subdomain', {
      username: this.u(input.username),
      ...compact({
        organization: input.organization,
        subdomain: input.subdomain,
        base_domain: input.baseDomain,
        hosted_zone_id: input.hostedZoneId,
        target_ip: input.targetIp,
        target_cname: input.targetCname,
        record_type: input.recordType,
        ttl: input.ttl,
      }),
    });
  }

  /** Attach instances to an application load balancer. POST /api/v2/connector/lb/attach */
  async attachLoadBalancer(input: ConnectorLBInput): Promise<Result> {
    if (!input.name) throw new Error('connector.attachLoadBalancer: name is required');
    if (!input.instanceIds?.length) throw new Error('connector.attachLoadBalancer: instanceIds is required');
    return this.post('/api/v2/connector/lb/attach', {
      username: this.u(input.username),
      instance_ids: input.instanceIds,
      ...compact({
        organization: input.organization,
        name: input.name,
        port: input.port,
        health_path: input.healthPath,
        region: input.region,
        subdomain: input.subdomain,
        vpc_id: input.vpcId,
        subnet_ids: input.subnetIds,
      }),
    });
  }
}

/** Drop undefined / null / empty-string fields so the node applies its own defaults. */
function compact(o: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(o)) {
    if (v === undefined || v === null || v === '') continue;
    out[k] = v;
  }
  return out;
}

File: connector/connector.go
// Package connector is the resource module for SDK-direct cloud deploys.
//
// Unlike the cloud package — which provisions via server-side Terraform under
// /api/v2/tenant/provision/* — connector targets the tenant node's
// /api/v2/connector/* surface, which provisions by calling cloud-provider SDKs
// and REST APIs directly: no Terraform, no Bash, no state files. It covers:
//
//   - EC2 VM provision / studio deploy (SCP+SSH) / terminate
//   - S3 static-site deploy and bucket creation
//   - GCP Cloud Run container deploy (Cloud Run Admin REST v2)
//   - Route53 subdomain records
//   - ALB attach with target group
//
// Every request carries a username for the Vault-backed credential lookup the
// node performs (connectors.FetchCloudCredentials). When a request leaves
// Username empty the Client fills it from AuthedUsername.
package connector

import (
	"context"

	"github.com/prodxcloud/vxcloud/transport"
)

// Client is the connector facade. Acquire it from the SDK with c.Connector().
type Client struct {
	T              *transport.Transport
	NodeURL        string
	AuthedUsername string
}

// ── Request types (JSON tags mirror services/connectors/connector_types.go) ──

// VMRequest provisions a single EC2 instance via the EC2 SDK.
type VMRequest struct {
	Username     string            `json:"username"`
	Organization string            `json:"organization,omitempty"`
	SessionID    string            `json:"session_id,omitempty"`
	InstanceName string            `json:"instance_name"`
	InstanceType string            `json:"instance_type,omitempty"`
	Region       string            `json:"region,omitempty"`
	OS           string            `json:"os,omitempty"`
	VolumeSize   int32             `json:"volume_size,omitempty"`
	KeyPairName  string            `json:"key_pair_name,omitempty"`
	Tags         map[string]string `json:"tags,omitempty"`
}

// DeployToVMRequest provisions an EC2 instance then pushes a studio project to
// it over SFTP and optionally runs a deploy command.
type DeployToVMRequest struct {
	Username     string            `json:"username"`
	Organization string            `json:"organization,omitempty"`
	SessionID    string            `json:"session_id"`
	StoragePath  string            `json:"storage_path,omitempty"`
	InstanceName string            `json:"instance_name,omitempty"`
	InstanceType string            `json:"instance_type,omitempty"`
	Region       string            `json:"region,omitempty"`
	OS           string            `json:"os,omitempty"`
	VolumeSize   int32             `json:"volume_size,omitempty"`
	KeyPairName  string            `json:"key_pair_name,omitempty"`
	Subdomain    string            `json:"subdomain,omitempty"`
	DeployCmd    string            `json:"deploy_command,omitempty"`
	Tags         map[string]string `json:"tags,omitempty"`
}

// StaticSiteRequest deploys a studio project to S3 as a static website.
type StaticSiteRequest struct {
	Username      string `json:"username"`
	Organization  string `json:"organization,omitempty"`
	SessionID     string `json:"session_id"`
	StoragePath   string `json:"storage_path,omitempty"`
	BucketName    string `json:"bucket_name,omitempty"`
	Region        string `json:"region,omitempty"`
	IndexDocument string `json:"index_document,omitempty"`
	ErrorDocument string `json:"error_document,omitempty"`
	Subdomain     string `json:"subdomain,omitempty"`
	CustomDomain  string `json:"custom_domain,omitempty"`
}

// BucketRequest creates a single S3 bucket.
type BucketRequest struct {
	Username     string `json:"username"`
	Organization string `json:"organization,omitempty"`
	BucketName   string `json:"bucket_name"`
	Region       string `json:"region,omitempty"`
	Public       bool   `json:"public,omitempty"`
}

// CloudRunRequest deploys a container image to Google Cloud Run.
type CloudRunRequest struct {
	Username     string            `json:"username"`
	Organization string            `json:"organization,omitempty"`
	SessionID    string            `json:"session_id,omitempty"`
	ServiceName  string            `json:"service_name"`
	Region       string            `json:"region,omitempty"`
	Image        string            `json:"image"`
	Port         int32             `json:"port,omitempty"`
	EnvVars      map[string]string `json:"env_vars,omitempty"`
	Memory       string            `json:"memory,omitempty"`
	CPU          string            `json:"cpu,omitempty"`
	MaxInstances int32             `json:"max_instances,omitempty"`
	Subdomain    string            `json:"subdomain,omitempty"`
	AllowPublic  bool              `json:"allow_public,omitempty"`
}

// SubdomainRequest creates a Route53 DNS record.
type SubdomainRequest struct {
	Username     string `json:"username"`
	Organization string `json:"organization,omitempty"`
	Subdomain    string `json:"subdomain"`
	BaseDomain   string `json:"base_domain,omitempty"`
	HostedZoneID string `json:"hosted_zone_id,omitempty"`
	TargetIP     string `json:"target_ip,omitempty"`
	TargetCNAME  string `json:"target_cname,omitempty"`
	RecordType   string `json:"record_type,omitempty"`
	TTL          int64  `json:"ttl,omitempty"`
}

// LBRequest attaches instances to an application load balancer.
type LBRequest struct {
	Username     string   `json:"username"`
	Organization string   `json:"organization,omitempty"`
	Name         string   `json:"name"`
	InstanceIDs  []string `json:"instance_ids"`
	Port         int64    `json:"port,omitempty"`
	HealthPath   string   `json:"health_path,omitempty"`
	Region       string   `json:"region,omitempty"`
	Subdomain    string   `json:"subdomain,omitempty"`
	VPCID        string   `json:"vpc_id,omitempty"`
	SubnetIDs    []string `json:"subnet_ids,omitempty"`
}

// ── Result types ────────────────────────────────────────────────────────────

// VMResult is returned after provisioning a VM.
type VMResult struct {
	InstanceID string `json:"instance_id"`
	PublicIP   string `json:"public_ip"`
	PrivateIP  string `json:"private_ip"`
	State      string `json:"state"`
	Region     string `json:"region"`
	Provider   string `json:"provider"`
}

// BucketResult is returned after creating a bucket.
type BucketResult struct {
	BucketName string `json:"bucket_name"`
	Region     string `json:"region"`
	ARN        string `json:"arn"`
	Provider   string `json:"provider"`
}

// DeployResult is returned for composite deploy operations (VM/S3).
type DeployResult struct {
	Status     string `json:"status"`
	Message    string `json:"message"`
	AccessURL  string `json:"access_url"`
	Provider   string `json:"provider"`
	Region     string `json:"region"`
	InstanceID string `json:"instance_id,omitempty"`
	PublicIP   string `json:"public_ip,omitempty"`
	BucketName string `json:"bucket_name,omitempty"`
	WebsiteURL string `json:"website_url,omitempty"`
	Subdomain  string `json:"subdomain,omitempty"`
	FQDN       string `json:"fqdn,omitempty"`
}

// SubdomainResult is returned after creating a DNS record.
type SubdomainResult struct {
	FQDN       string `json:"fqdn"`
	RecordType string `json:"record_type"`
	Target     string `json:"target"`
	TTL        int64  `json:"ttl"`
}

// LBResult is returned after creating/attaching a load balancer.
type LBResult struct {
	LoadBalancerARN string `json:"load_balancer_arn"`
	DNSName         string `json:"dns_name"`
	TargetGroupARN  string `json:"target_group_arn"`
	FQDN            string `json:"fqdn,omitempty"`
}

// CloudRunResult is returned after deploying to Cloud Run.
type CloudRunResult struct {
	ServiceName string `json:"service_name"`
	ServiceURL  string `json:"service_url"`
	Region      string `json:"region"`
	ProjectID   string `json:"project_id"`
	FQDN        string `json:"fqdn,omitempty"`
}

// ── Methods ─────────────────────────────────────────────────────────────────

// ProvisionVM launches a single EC2 instance (no file deploy).
// POST /api/v2/connector/vm/provision
func (c *Client) ProvisionVM(ctx context.Context, req VMRequest) (*VMResult, error) {
	req.Username = c.user(req.Username)
	var out VMResult
	if err := c.post(ctx, "connector.ProvisionVM", "/api/v2/connector/vm/provision", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// DeployToVM provisions an EC2 instance and pushes a studio project to it.
// POST /api/v2/connector/vm/studio/deploy
func (c *Client) DeployToVM(ctx context.Context, req DeployToVMRequest) (*DeployResult, error) {
	req.Username = c.user(req.Username)
	var out DeployResult
	if err := c.post(ctx, "connector.DeployToVM", "/api/v2/connector/vm/studio/deploy", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// TerminateVM terminates an EC2 instance.
// POST /api/v2/connector/vm/terminate
func (c *Client) TerminateVM(ctx context.Context, instanceID, organization string) error {
	body := map[string]string{
		"username":     c.user(""),
		"organization": organization,
		"instance_id":  instanceID,
	}
	var out map[string]any
	return c.post(ctx, "connector.TerminateVM", "/api/v2/connector/vm/terminate", body, &out)
}

// CreateBucket creates a single S3 bucket.
// POST /api/v2/connector/s3/bucket
func (c *Client) CreateBucket(ctx context.Context, req BucketRequest) (*BucketResult, error) {
	req.Username = c.user(req.Username)
	var out BucketResult
	if err := c.post(ctx, "connector.CreateBucket", "/api/v2/connector/s3/bucket", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// DeployStaticSite deploys a studio project to S3 as a static website.
// POST /api/v2/connector/s3/studio/deploy
func (c *Client) DeployStaticSite(ctx context.Context, req StaticSiteRequest) (*DeployResult, error) {
	req.Username = c.user(req.Username)
	var out DeployResult
	if err := c.post(ctx, "connector.DeployStaticSite", "/api/v2/connector/s3/studio/deploy", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// DeployCloudRun deploys a container image to Google Cloud Run.
// POST /api/v2/connector/gcr/studio/deploy
func (c *Client) DeployCloudRun(ctx context.Context, req CloudRunRequest) (*CloudRunResult, error) {
	req.Username = c.user(req.Username)
	var out CloudRunResult
	if err := c.post(ctx, "connector.DeployCloudRun", "/api/v2/connector/gcr/studio/deploy", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// CreateSubdomain creates a Route53 DNS record.
// POST /api/v2/connector/dns/subdomain
func (c *Client) CreateSubdomain(ctx context.Context, req SubdomainRequest) (*SubdomainResult, error) {
	req.Username = c.user(req.Username)
	var out SubdomainResult
	if err := c.post(ctx, "connector.CreateSubdomain", "/api/v2/connector/dns/subdomain", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// AttachLoadBalancer creates an ALB + target group and registers instances.
// POST /api/v2/connector/lb/attach
func (c *Client) AttachLoadBalancer(ctx context.Context, req LBRequest) (*LBResult, error) {
	req.Username = c.user(req.Username)
	var out LBResult
	if err := c.post(ctx, "connector.AttachLoadBalancer", "/api/v2/connector/lb/attach", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ── internals ───────────────────────────────────────────────────────────────

func (c *Client) post(ctx context.Context, op, path string, body, out any) error {
	u := transport.JoinURL(c.NodeURL, path)
	return c.T.JSON(ctx, op, "POST", u, body, out)
}

func (c *Client) user(requested string) string {
	if requested != "" {
		return requested
	}
	return c.AuthedUsername
}
